### PR TITLE
Extract parquet statistics from f16 columns, add `ScalarValue::Float16`

### LINKED ISSUE #10757: Extract parquet statistics from `f16` columns
### Is your feature request related to a problem or challenge?

Part of https://github.com/apache/datafusion/issues/10453, where we are filling out support for extracting statistics for all data types from parquet files

At the moment, even if statistics are extracted for a different type (like `Int32`) the PruningPredicate will attempt to cast these values to the correct type:

https://github.com/apache/datafusion/blob/acd7106fa40fad58f50ae06227971c51073d8f48/datafusion/core/src/physical_optimizer/pruning.rs#L909-L911

However, in order to be efficient and ensure the cast kernel doesn't add anything incorrectly, we should be extracting the parquet statistics as the correct Array type directly. It turns out we do not do this yet for several types and those types do not have good (or any) test coverage. We almost missed this in https://github.com/apache/datafusion/pull/10711 in @xinlifoobar 

Thus, we need to add support and tests for other types

### Describe the solution you'd like

1. Add a new test to [arrow_stastics.rs](https://github.com/apache/datafusion/blob/main/datafusion/core/tests/parquet/arrow_statistics.rs) (run this with `cargo test --test parquet_exec`) with the relevant type
2. Potentially add a new case to the match here https://github.com/apache/datafusion/blob/acd7106fa40fad58f50ae06227971c51073d8f48/datafusion/core/src/datasource/physical_plan/parquet/statistics.rs#L61-L182  to get the correct types

Here are some example PRs: 
1. https://github.com/apache/datafusion/pull/10729
2. https://github.com/apache/datafusion/pull/10593

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Lordworms**
## Which issue does this PR close?

Closes #10757

### CODE REVIEW

File: datafusion/common/src/scalar/mod.rs [hunk 1]
```diff
@@ -1700,7 +1722,7 @@ impl ScalarValue {
                 );
             }
         };
-
+        println!("array is {:?}", array);
```
Comment (alamb) [inline datafusion/common/src/scalar/mod.rs:1725]: I think this should be removed prior to merging

```suggestion
```

**Review by alamb — COMMENTED** [review]
Thank you @Lordworms  -- other than the `println` I think this PR is ready to go

**Comment by Lordworms** [comment]
seems like there is a CI/CD issue

**Review by alamb — APPROVED** [review]
Thank you @Lordworms  -- I took the liberty of merging up from main to resolve some conflicts in this PR

**Pull request merged**